pkg/ddl/notifier/subscribe.go

```diff
@@ -0,0 +1,208 @@
+// ddlNotifier is started. If the handler can't immediately serve the handling
+// after registering, it can return nil to tell the ddlNotifier to act like the
+// change has been handled, or return ErrNotReadyRetryLater to hold the change
+// and re-handle later.
+type SchemaChangeHandler func(
+	ctx context.Context,
+	sctx sessionctx.Context,
+	change *SchemaChangeEvent,
+) error
+
+// ErrNotReadyRetryLater should be returned by a registered handler that is not
+// ready to process the events.
+var ErrNotReadyRetryLater = errors.New("not ready, retry later")
+
+// HandlerID is the type of the persistent ID used to register a handler. Every
+// ID occupies a bit in a BIGINT column, so at most we can only have 64 IDs. To
+// avoid duplicate IDs, all IDs should be defined in below declaration.
+type HandlerID int
+
+const (
+	// TestHandlerID is used for testing only.
+	TestHandlerID HandlerID = iota
+)
+
+// RegisterHandler must be called with an exclusive and fixed HandlerID for each
+// handler to register the handler. Illegal ID will panic. RegisterHandler should
+// not be called after the global ddlNotifier is started.
+//
+// RegisterHandler is not concurrency-safe.
+func RegisterHandler(id HandlerID, handler SchemaChangeHandler) {
+	intID := int(id)
+	if intID < 0 || intID >= 64 {
+		panic(fmt.Sprintf("illegal HandlerID: %d", id))
+	}
+
+	if _, ok := globalDDLNotifier.handlers[id]; ok {
+		panic(fmt.Sprintf("HandlerID %d already registered", id))
+	}
+	globalDDLNotifier.handlers[id] = handler
+}
+
+type ddlNotifier struct {
+	ownedSCtx    sessionctx.Context
+	store        Store
+	handlers     map[HandlerID]SchemaChangeHandler
+	pollInterval time.Duration
+}
+
+var globalDDLNotifier *ddlNotifier
+
+// InitDDLNotifier initializes the global ddlNotifier. It should be called only
+// once and before any RegisterHandler call. The ownership of the sctx is passed
+// to the ddlNotifier.
+func InitDDLNotifier(
+	sctx sessionctx.Context,
+	store Store,
+	pollInterval time.Duration,
+) {
+	globalDDLNotifier = &ddlNotifier{
+		ownedSCtx:    sctx,
+		store:        store,
+		handlers:     make(map[HandlerID]SchemaChangeHandler),
+		pollInterval: pollInterval,
+	}
+}
+
+// ResetDDLNotifier is used for testing only.
+func ResetDDLNotifier() { globalDDLNotifier = nil }
+
+// StartDDLNotifier starts the global ddlNotifier. It will block until the
+// context is canceled.
+func StartDDLNotifier(ctx context.Context) {
+	globalDDLNotifier.Start(ctx)
+}
+
+// Start starts the ddlNotifier. It will block until the context is canceled.
+func (n *ddlNotifier) Start(ctx context.Context) {
+	ctx = kv.WithInternalSourceType(ctx, kv.InternalDDLNotifier)
+	ticker := time.NewTicker(n.pollInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := n.processEvents(ctx); err != nil {
+				log.FromContext(ctx).Error("Error processing events", zap.Error(err))
```

Comment:
This logger seems like the `DDLIngestLogger`, here `DDLLogger`  looks more accurate
```
[2024/09/20 16:52:54.168 +08:00] [ERROR] [subscribe.go:151] ["Error processing event"] **[category=ddl-ingest]** [ddlJobID=3] [multiSchemaChangeSeq=-1] [handlerID=0] [error=EOF] [errorVerbose="EOF\ngithub.com/pingcap/errors.AddStack\n\t/Users/fanzhou/.gvm/pkgsets/go1.22.4/global/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/errors.go:178\ngithub.com/pingcap/errors.Trace\n\t/Users/fanzhou/.gvm/pkgsets/go1.22.4/global/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/ddl/notifier.(*ddlNotifier).processEventForHandler\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:191\ngithub.com/pingcap/tidb/pkg/ddl/notifier.(*ddlNotifier).processEvents\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:147\ngithub.com/pingcap/tidb/pkg/ddl/notifier.(*ddlNotifier).Start\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:126\ngithub.com/pingcap/tidb/pkg/ddl/notifier.StartDDLNotifier\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/subscribe.go:113\ngithub.com/pingcap/tidb/pkg/ddl/notifier_test.TestBasicPubSub.func2\n\t/Users/fanzhou/Documents/GitHub/tidb/pkg/ddl/notifier/testkit_test.go:105\nruntime.goexit\n\t/Users/fanzhou/.gvm/gos/go1.22.4/src/runtime/asm_arm64.s:1222"]

```

Reply:
https://github.com/pingcap/tidb/pull/56182/commits/734a9c580fca6bb8dfba25dbd8316a19d2b1e5ed